libbeat/common/event.go

```diff
@@ -1,92 +1,223 @@
 package common
 
 import (
+	"encoding"
 	"encoding/json"
+	"fmt"
 	"reflect"
-	"time"
+	"strconv"
+	"strings"
 
 	"github.com/elastic/beats/libbeat/logp"
+
+	"github.com/pkg/errors"
 )
 
-func MarshallUnmarshall(v interface{}) (MapStr, error) {
-	// decode and encode JSON
-	marshaled, err := json.Marshal(v)
-	if err != nil {
-		logp.Warn("marshal err: %v", err)
-		return nil, err
+const eventDebugSelector = "event"
+
+var eventDebugf = logp.MakeDebug(eventDebugSelector)
+
+var textMarshalerType = reflect.TypeOf((*encoding.TextMarshaler)(nil)).Elem()
+
+// ConvertToGenericEvent normalizes the types contained in the given MapStr.
+//
+// Nil values in maps are dropped during the conversion. Any unsupported types
+// that are found in the MapStr are dropped and warnings are logged.
+func ConvertToGenericEvent(m MapStr) MapStr {
+	event, errs := normalizeMap("", m)
+	if len(errs) > 0 {
+		logp.Warn("Unsuccessful conversion to generic event: %v errors: %v, "+
```

Comment:
I'm thinking instead of logging, it should return the error to the caller. Though, I'm on on the fence about this. Originally this method logged the issues it detected and still passed back a `MapStr` to the caller. It had weak guarantees about the resulting MapStr being "generic".

I'm thinking the method signature should be `func ConvertToGenericEvent(m MapStr) (MapStr, error)` and on error we return a `nil` MapStr. This means that if you write a bad event, it will be dropped (but logged).